File: services/ai/src/rag/ingest.py
"""Document ingestion pipeline - chunk, embed, and store in Qdrant."""

import uuid
from collections.abc import Callable

import structlog

from src.adapters.embedding import EmbeddingClient
from src.adapters.qdrant import QdrantAdapter
from src.rag.structured_chunker import chunk_document

log = structlog.get_logger()


async def ingest_document(
    text: str,
    source_id: str,
    source_name: str,
    embedding_client: EmbeddingClient,
    metadata: dict[str, str] | None = None,
    qdrant: QdrantAdapter | None = None,
    mime_type: str = "text/plain",
    filename: str = "",
    heartbeat_fn: Callable[[str], None] | None = None,
) -> int:
    """Ingest a document into the knowledge base.

    Pipeline: text → structured chunk → embed → upsert to Qdrant.

    Args:
        text: Document text content
        source_id: Unique identifier for the source document
        source_name: Human-readable source name
        embedding_client: Embedding client (requires API key)
        metadata: Additional metadata (tenant_id, course_id, team_id, session_id)
        qdrant: Optional Qdrant adapter
        mime_type: MIME type for parser selection
        filename: Filename for parser selection fallback
        heartbeat_fn: Optional heartbeat callback for Temporal activity

    Returns:
        Number of chunks ingested
    """
    qdrant = qdrant or QdrantAdapter()
    metadata = metadata or {}

    # Step 1: Structure-aware chunking
    if heartbeat_fn:
        heartbeat_fn("chunking document")
    structured_chunks = chunk_document(text, mime_type=mime_type, filename=filename or source_name)
    if not structured_chunks:
        log.warn("no chunks produced from document", source_id=source_id)
        return 0

    log.info(
        "chunked document",
        source_id=source_id,
        source_name=source_name,
        chunk_count=len(structured_chunks),
    )

    # Step 2: Generate embeddings
    if heartbeat_fn:
        heartbeat_fn(f"embedding {len(structured_chunks)} chunks")
    chunk_texts = [c.content for c in structured_chunks]
    embeddings = await embedding_client.embed(chunk_texts)

    # Step 3: Build Qdrant points with section heading metadata
    points = []
    for i, (sc, embedding) in enumerate(zip(structured_chunks, embeddings)):
        point_id = str(uuid.uuid5(uuid.NAMESPACE_URL, f"{source_id}:{i}"))
        payload = {
            "content": sc.content,
            "source_id": source_id,
            "source_name": source_name,
            "chunk_index": i,
            "section_heading": sc.section_heading,
            **metadata,
        }
        points.append(
            {
                "id": point_id,
                "vector": embedding,
                "payload": payload,
            }
        )

        # Heartbeat every 50 chunks during point building
        if heartbeat_fn and i > 0 and i % 50 == 0:
            heartbeat_fn(f"prepared {i}/{len(structured_chunks)} chunks")

    # Step 4: Upsert to Qdrant in batches
    if heartbeat_fn:
        heartbeat_fn("storing in vector database")
    await qdrant.ensure_collection()
    await qdrant.upsert_batch(points, batch_size=100)

    log.info(
        "document ingested",
        source_id=source_id,
        chunks=len(structured_chunks),
    )

    return len(structured_chunks)

File: backend/internal/application/workflow/activities/ingest.go
package activities

import (
	"context"
	"fmt"

	"github.com/google/uuid"
	"go.temporal.io/sdk/activity"

	"github.com/sogos/mirai-backend/internal/infrastructure/external/vectordb"
	"github.com/sogos/mirai-backend/internal/infrastructure/rag"
)

// IngestDocumentInput is the input for the IngestDocument activity.
type IngestDocumentInput struct {
	Text       string            `json:"text"`
	SourceID   string            `json:"source_id"`
	SourceName string            `json:"source_name"`
	APIKey     string            `json:"api_key"`
	MimeType   string            `json:"mime_type"`
	Filename   string            `json:"filename"`
	Metadata   map[string]string `json:"metadata"`
}

// IngestDocumentOutput is the output from the IngestDocument activity.
type IngestDocumentOutput struct {
	ChunkCount int32 `json:"chunk_count"`
	TokenCount int32 `json:"token_count"`
}

// IngestDocument chunks a document, generates embeddings via Gemini, and upserts to Qdrant.
func (a *GoActivities) IngestDocument(ctx context.Context, input IngestDocumentInput) (*IngestDocumentOutput, error) {
	activity.RecordHeartbeat(ctx, "starting ingestion")
	logger := activity.GetLogger(ctx)

	if a.EmbeddingClient == nil {
		return nil, fmt.Errorf("embedding client not configured")
	}
	if a.QdrantClient == nil {
		return nil, fmt.Errorf("qdrant client not configured")
	}

	// Step 1: Structure-aware chunking
	activity.RecordHeartbeat(ctx, "chunking document")
	chunks := rag.ChunkDocument(input.Text, input.MimeType, input.Filename)
	if len(chunks) == 0 {
		logger.Warn("no chunks produced from document", "sourceID", input.SourceID)
		return &IngestDocumentOutput{ChunkCount: 0, TokenCount: 0}, nil
	}

	logger.Info("chunked document",
		"sourceID", input.SourceID,
		"sourceName", input.SourceName,
		"chunkCount", len(chunks),
	)

	// Step 2: Generate embeddings in batches with heartbeats
	batchSize := a.EmbeddingClient.BatchSize()
	allEmbeddings := make([][]float32, 0, len(chunks))

	for batchStart := 0; batchStart < len(chunks); batchStart += batchSize {
		batchEnd := batchStart + batchSize
		if batchEnd > len(chunks) {
			batchEnd = len(chunks)
		}

		activity.RecordHeartbeat(ctx, fmt.Sprintf("embedding chunks %d-%d/%d", batchStart+1, batchEnd, len(chunks)))

		texts := make([]string, batchEnd-batchStart)
		for i, chunk := range chunks[batchStart:batchEnd] {
			texts[i] = chunk.Content
		}

		embeddings, err := a.EmbeddingClient.EmbedDocuments(ctx, input.APIKey, texts)
		if err != nil {
			return nil, fmt.Errorf("embed batch %d-%d: %w", batchStart, batchEnd, err)
		}
		allEmbeddings = append(allEmbeddings, embeddings...)
	}

	// Step 3: Build Qdrant points with UUID5 IDs
	points := make([]vectordb.Point, len(chunks))
	for i, chunk := range chunks {
		pointID := uuid.NewSHA1(uuid.NameSpaceURL, []byte(fmt.Sprintf("%s:%d", input.SourceID, i)))

		payload := map[string]interface{}{
			"content":         chunk.Content,
			"source_id":       input.SourceID,
			"source_name":     input.SourceName,
			"chunk_index":     i,
			"section_heading": chunk.SectionHeading,
		}
		for k, v := range input.Metadata {
			payload[k] = v
		}

		points[i] = vectordb.Point{
			ID:      pointID.String(),
			Vector:  allEmbeddings[i],
			Payload: payload,
		}

		if i > 0 && i%50 == 0 {
			activity.RecordHeartbeat(ctx, fmt.Sprintf("prepared %d/%d chunks", i, len(chunks)))
		}
	}

	// Step 4: Upsert to Qdrant in batches of 100
	activity.RecordHeartbeat(ctx, "storing in vector database")

	if err := a.QdrantClient.EnsureCollection(ctx, a.QdrantCollection, len(allEmbeddings[0])); err != nil {
		return nil, fmt.Errorf("ensure collection: %w", err)
	}

	upsertBatchSize := 100
	for batchStart := 0; batchStart < len(points); batchStart += upsertBatchSize {
		batchEnd := batchStart + upsertBatchSize
		if batchEnd > len(points) {
			batchEnd = len(points)
		}

		if err := a.QdrantClient.Upsert(ctx, a.QdrantCollection, points[batchStart:batchEnd]); err != nil {
			return nil, fmt.Errorf("upsert batch %d-%d: %w", batchStart, batchEnd, err)
		}

		activity.RecordHeartbeat(ctx, fmt.Sprintf("upserted %d/%d points", batchEnd, len(points)))
	}

	// Estimate token count (chars / 4 is a reasonable approximation)
	tokenCount := int32(len(input.Text) / 4)

	logger.Info("document ingested",
		"sourceID", input.SourceID,
		"chunks", len(chunks),
		"tokenCount", tokenCount,
	)

	return &IngestDocumentOutput{
		ChunkCount: int32(len(chunks)),
		TokenCount: tokenCount,
	}, nil
}
